util/progress/progressui/display.go

```diff
@@ -265,6 +274,140 @@ func (d *plainDisplay) done() {
 	d.t.printErrorLogs(d.t.w)
 }
 
+type jsonDisplay struct {
+	t    *trace
+	logs *logMux
+	enc  *json.Encoder
+}
+
+// newJSONDisplay creates a new SolveStatusDisplay that outputs the completed vertex
+// progress indicator for each vertex in JSON format.
+func newJSONDisplay(w io.Writer, opts ...DisplayOpt) Display {
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	return Display{
+		disp: &jsonDisplay{
+			t:    newTrace(w, false),
+			logs: newLogMux(),
+			enc:  enc,
+		},
+	}
+}
+
+func (d *jsonDisplay) init(displayLimiter *rate.Limiter) {
+	// Display limiter is ignored for the JSON display.
+}
+
+func (d *jsonDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, 80)
+
+	// JSON format keeps the logs separated into their streams
+	// and takes the aggregated result. The trace does not do that
+	// so we use and update logMux for log messages by digest/stream.
```

Comment:
Do we need to disable log behavior in `d.t` to avoid duplicates?

Reply:
The implementation should just ignore the logs. The `trace` will still do the log splitting and aggregating stdout/stderr into the same stream, but this specific display just ignores that property. It would be possible for us to disable the log behavior to improve performance, but it would make that section of code more complicated and I think that section of code may already be a bit too complicated and in need of a refactor so, unless it's causing a documented performance problem, I think we should just ignore it.

It won't create a wrong answer.

Reply:
If it means that we use double the memory then we should fix it. But if we don't need to store everything for JSON and can release the memory once vertex is printed then not needed until some benchmark proves that this is a bottleneck.